app/com/linkedin/drelephant/tuning/AbstractBaselineManager.java

```diff
@@ -0,0 +1,145 @@
+          + "(SELECT job_exec_id, SUM(resource_used/(1024 * 3600)) AS resource_used, "
+          + "SUM((finish_time - start_time - total_delay)/(1000 * 60))  AS execution_time, "
+          + "MAX(start_time) AS start_time " + "FROM yarn_app_result WHERE job_def_id=:jobDefId "
+          + "GROUP BY job_exec_id " + "ORDER BY start_time DESC " + "LIMIT :num) temp";
+  protected String avgInputSizeSQL = "SELECT AVG(inputSizeInBytes) as avgInputSizeInMB FROM "
+      + "(SELECT job_exec_id, SUM(cast(value as decimal)) inputSizeInBytes, MAX(start_time) AS start_time "
+      + "FROM yarn_app_result yar INNER JOIN yarn_app_heuristic_result yahr " + "ON yar.id=yahr.yarn_app_result_id "
+      + "INNER JOIN yarn_app_heuristic_result_details yahrd " + "ON yahr.id=yahrd.yarn_app_heuristic_result_id "
+      + "WHERE job_def_id=:jobDefId AND yahr.heuristic_name='" + CommonConstantsHeuristic.MAPPER_SPEED + "' "
+      + "AND yahrd.name='" + CommonConstantsHeuristic.TOTAL_INPUT_SIZE_IN_MB + "' "
+      + "GROUP BY job_exec_id ORDER BY start_time DESC LIMIT :num ) temp";
+  private final Logger logger = Logger.getLogger(getClass());
+  protected Integer _numJobsForBaseline = null;
+
+  /*
+  Execute the whole logic for Base line Computing
+  1) Get the jobs for which Baseline have to be calculated
+  2) Calculate the baseline
+  3) Update the base line
+  4) Update the metrics .
+   */
+  @Override
+  public final Boolean execute() {
+    logger.info("Executing BaseLine");
+    Boolean baseLineComputationDone = false, databaseUpdateDone = false, updateMetricsDone = false;
+    List<TuningJobDefinition> tuningJobDefinitions = detectJobsForBaseLineComputation();
+    if (tuningJobDefinitions != null && tuningJobDefinitions.size() >= 1) {
+      logger.info("Computing BaseLine");
+      baseLineComputationDone = calculateBaseLine(tuningJobDefinitions);
+    }
+    if (baseLineComputationDone) {
+      logger.info("Updating Database");
+      databaseUpdateDone = updateDataBase(tuningJobDefinitions);
+    }
+    if (databaseUpdateDone) {
+      logger.info("Updating Metrics");
+      updateMetricsDone = updateMetrics(tuningJobDefinitions);
+    }
+    logger.info("Baseline Done");
+    return updateMetricsDone;
+  }
+
+  /**
+   * Fetches the jobs whose baseline is to be computed.
+   * This is done by returning the jobs with null average resource usage
+   * @return List of jobs whose baseline needs to be added
+   */
+
+  protected abstract List<TuningJobDefinition> detectJobsForBaseLineComputation();
+
+  /**
+   * Adds baseline metric values for a job
+   * @param tuningJobDefinitions Job for which baseline is to be computed
+   */
+
+  protected Boolean calculateBaseLine(List<TuningJobDefinition> tuningJobDefinitions) {
+    for (TuningJobDefinition tuningJobDefinition : tuningJobDefinitions) {
+      try {
+        logger.info("Computing and updating baseline metric values for job: " + tuningJobDefinition.job.jobName);
```

Comment:
tuning_job_definition has a column named tuningAlgorithm which is meant for PSO/IPSO. While with only two tuning types i.e. HBT/OBT as of now, we can probably use this table for HBT when tuningAlgorithm is NULL. But we should ideally have a column for tuning type to make the code extensible for a new tuning type in future (say ML based).Same goes for other tables as well